### DISCUSSION TITLE
Widget new task behaviour

### DISCUSSION THREAD

**Discussion by Tirpitz93** [TechbeeAT/jtxBoard#219] [Ideas]
I believe the add new button on the widget should default to the list selected in the widget (especially if only one list is selected).
Also when using the add new from widget I would like to have a save and add new.

When I use the widget I usually want to quickly add items to my groceries list or chores list.

**Comment by patrickunterwegs** [comment]
Hey @Tirpitz93 !
Thanks for your remarks!

> I believe the add new button on the widget should default to the list selected in the widget (especially if only one list is selected).

You're right and indeed I checked this not so long ago, but that simple thing was driving me so crazy during implementation, that I stopped it and put it aside 😅   -> #223 

> Also when using the add new from widget I would like to have a save and add new.

I'll check in #222

**Reply by patrickunterwegs** [reply]
Hey @Tirpitz93 , both issues are solved now, if you'd like to make an early check you can either join the beta-program on Google Play and download 2.02.01-rc08.gplay or you take the apk from the release section here on GitHub. Otherwise the new version will be available probably around the end of next week!

**Reply by Tirpitz93** [reply]
Absolute legend.
I'm stuck on rc07 via gplay

**Reply by patrickunterwegs** [reply]
Sorry, I forgot to say that I just uploaded it, Google takes a few hours to approve it, so it should come some time during the day 😉

**Comment by Tirpitz93** [comment]
Tested [v2.2.1-rc08](https://github.com/TechbeeAT/jtxBoard/releases/tag/v2.2.1-rc08) 

- Widget's new button does select the selected collection
-  the save & new just seems to scroll to the top of the list view in the main app.

**Reply by patrickunterwegs** [reply]
Thanks @Tirpitz93 !
Regarding Save & new: are you sure you don't have a filter activated that immediately filters the new entry?
Or do you maybe have a grouping enabled?

**Reply by Tirpitz93** [reply]
just waiting for my screenshots to upload...
| ###### | ######## |
|----------|----------|
| ![Screenshot_20230124-131821](https://user-images.githubusercontent.com/6187560/214315653-e89be972-d05b-4332-8e22-1fc82303057c.jpg) | ![Screenshot_20230124-131738](https://user-images.githubusercontent.com/6187560/214315612-aa526937-0002-4afd-a352-b26b18846521.jpg) |
| ![Screenshot_20230124-131757](https://user-images.githubusercontent.com/6187560/214315619-97a7e927-e007-4c0d-9fc4-1e36cd3b85db.jpg) | ![Screenshot_20230124-131813](https://user-images.githubusercontent.com/6187560/214315631-d296a7f1-e50a-4433-b220-1a6b41802512.jpg) |
|----------|------------|

**Reply by patrickunterwegs** [reply]
Ah, I think I understand. You mean the entry is added, but the dialog is closing although it should stay open, right?

In fact, I checked the behavior that the right tab gets opened, I checked the behavior when a new entry is added through the lightning icon out of the app, but I have overlooked that adding a new entry from the widget has a specific condition that caused the dialog to close. Give me a few minutes and I'll get it fixed 😉

**Reply by Tirpitz93** [reply]
Sorry, Typo in my reply, 
Widget's new button **doesnt** select the widgets filtered collection

**Reply by patrickunterwegs** [reply]
Well, that one was not in the specification 😄 I'm not sure if this is such a good idea. Currently the dialog should always take the collection that was last used. In the widget you can also filter multiple collections. Which one should be preset then? The first in the list? But then wouldn't it be better to keep the logic as is and preselect the last one that was used?

**Reply by Tirpitz93** [reply]
> Well, that one was not in the specification 😄 I'm not sure if this is such a good idea. Currently the dialog should always take the collection that was last used. In the widget you can also filter multiple collections. Which one should be preset then? The first in the list? But then wouldn't it be better to keep the logic as is and preselect the last one that was used?

I had it in the original message.
> _**I believe the add new button on the widget should default to the list selected in the widget (especially if only one list is selected).**_ Also when using the add new from widget I would like to have a save and add new.

Just not in the reply:
> Widget's new button does select the selected collection



Top of my head ideas:
 - If widget is filtered to a single collection, select this.
     else show additional dialog with list of collections shown in widget, then show create dialog. this would reduce the number of taps by one and reduce the chance that items end up in a random collection that is not shown. (or apply one of the options below)
- Select the first collection that is in the widgets filter.
- Select the LRU collection from the widget.

**Reply by patrickunterwegs** [reply]
> I believe the add new button on the widget should default to the list selected in the widget (especially if only one list is selected). Also when using the add new from widget I would like to have a save and add new.

The list is selected now. The collection is not selected.

**Reply by Tirpitz93** [reply]
> The list is selected now. The collection is not selected.

Sorry, in this context what is a list?

I was using them interchangeably

**Reply by patrickunterwegs** [reply]
It was just a misunderstanding, I was referring to the journals, notes or tasks list (-> the right tab should be selected when opening the app from the widget). You used the word list for collection.

**Reply by Tirpitz93** [reply]
> It was just a misunderstanding, I was referring to the journals, notes or tasks list (-> the right tab should be selected when opening the app from the widget). You used the word list for collection.

Yes it does that.

**Reply by patrickunterwegs** [reply]
> Top of my head ideas:
> 
>     If widget is filtered to a single collection, select this.
>     else show additional dialog with list of collections shown in widget, then show create dialog. this would reduce the number of taps by one and reduce the chance that items end up in a random collection that is not shown. (or apply one of the options below)
>     Select the first collection that is in the widgets filter.
>     Select the LRU collection from the widget.

I have started checking how this can be done best. Regrettably the widget framework makes it unnecessarily hard to achieve this...

**Comment by patrickunterwegs** [comment]
Hey @Tirpitz93 , 
about that one: 

>     If widget is filtered to a single collection, select this.
    else show additional dialog with list of collections shown in widget, then show create dialog. this would reduce the number of taps by one and reduce the chance that items end up in a random collection that is not shown. (or apply one of the options below)
    Select the first collection that is in the widgets filter.
    Select the LRU collection from the widget.

I have tried to implement it. The logic is: 
- Take the first out of the filtered collections from the widget, 
- if empty then take the last used collection, 
- if not possible take the first possible in the list

I'd appreciate if you could also test this a bit! I'll upload a new version 2.02.01-rc09 which should be approved by tomorrow!

I have been struggling with very weird behaviors when there are multiple widgets for the same module (journals, notes, tasks), then it was like the dialog showed randomly the right collection or not... I actually believe this is an error in the library provided by Google. I will do a bit more testing tomorrow if it's also reproducible on the device. But even if it is, but it still works correctly when there's only one widget per module, I'd publish it with the bug, as it either would have the advantage of selecting the right collection or just react as before.
What do you think?

**Reply by Tirpitz93** [reply]
I will install as soon as I see it on github.
I have switched to the manual install so I think that will be easier.
I have also started to install android IDE and will try and get the app to build. (at some point this week)

> I have tried to implement it. The logic is:
> 
>     * Take the first out of the filtered collections from the widget,
> 
>     * if empty then take the last used collection,
> 
>     * if not possible take the first possible in the list
> 
> 
> I'd appreciate if you could also test this a bit! I'll upload a new version 2.02.01-rc09 which should be approved by tomorrow!
> 
> I have been struggling with very weird behaviors when there are multiple widgets for the same module (journals, notes, tasks), then it was like the dialog showed randomly the right collection or not... I actually believe this is an error in the library provided by Google. I will do a bit more testing tomorrow if it's also reproducible on the device. But even if it is, but it still works correctly when there's only one widget per module, I'd publish it with the bug, as it either would have the advantage of selecting the right collection or just react as before. What do you think?

I like the sound of this.

**Reply by patrickunterwegs** [reply]
The build from source should be pretty straighforward with Android Studio, you should need any special config for a debug build 😉 
If you would like to use the apk from GitHub, I can recommend using the app Obtainium, it let's you configure the place of the apk on GitHub and will provide you with updates!

**Reply by Tirpitz93** [reply]
> The build from source should be pretty straighforward with Android Studio, you should need any special config for a debug build 😉 If you would like to use the apk from GitHub, I can recommend using the app Obtainium, it let's you configure the place of the apk on GitHub and will provide you with updates!
That app looks awesome!

**Reply by Tirpitz93** [reply]
> I have been struggling with very weird behaviors when there are multiple widgets for the same module (journals, notes, tasks), then it was like the dialog showed randomly the right collection or not... I actually believe this is an error in the library provided by Google. I will do a bit more testing tomorrow if it's also reproducible on the device. But even if it is, but it still works correctly when there's only one widget per module, I'd publish it with the bug, as it either would have the advantage of selecting the right collection or just react as before. What do you think?

My testing parameters:
- Tasks only
- widgets all on one home screen
- widgets left in 1x3 size
- each widget only has one collection selected.

I have a possible pattern:

- It selects the selected list from the most recently created widget.

- I dont think order on screen has anything to do with it.
- I can move widgets around and this doesnt affect the selected list.
- on deleting the most recently created widget selecting another add button selects the collection in the just deleted widget once, but then moves to the next most recently created widgets' collection.

**Reply by Tirpitz93** [reply]
I havent worked with android in a very long time, but it feels like a class vs instance variable/property issue.
I also dont know anything about the widgets API.

**Reply by patrickunterwegs** [reply]
Thank you for your observations!

> It selects the selected list from the most recently created widget.

This would also correspond to an observation that I made some time ago with another issue in the widget.
But I still believe it's a problem in the library. The code-snippet is more or less the same as when you're clicking on an entry and the specific entry is opening, the only difference is that the single entries are part of a "lazy" list.
The whole widget thing is a bit intransparent and I'm also not that experienced with all that connections. But the library is still in an alpha state, so I believe I could report this to Google (or ask on Stackoverflow). For the background, I used this new Glance library believing that Google would at least provide a Beta by the end of last year, so I thought it's better to use the new technology instead of using XML layouts for the widget. However, since then Google just provided another alpha-version and still there is no roadmap when they are going to release a beta or release version...

**Reply by patrickunterwegs** [reply]
I have opened an issue report in the Google issues: https://issuetracker.google.com/issues/266618721

**Comment by Tirpitz93** [comment]
While we are here discussing widget behaviour, how hard would it be to add a widget label field to the options that can be customized?

**Reply by patrickunterwegs** [reply]
I'm not sure what you mean by customized? What would such a label field do?

**Reply by Tirpitz93** [reply]
Show a custom label in the widget title instead of "Tasks" (as is the case for mine at the moment).
I have 3 or 4 on one page in the launcher and I would like to show a label on a per widget basis (as displaying the list of collections wouldnt work).

**Reply by patrickunterwegs** [reply]
That one is easy to do and already on the way 😉 See #232 / 2.02.01-rc10

**Reply by Tirpitz93** [reply]
> That one is easy to do and already on the way 😉 See #232 / 2.02.01-rc10

Works for me! 
Thanks

**Reply by Tirpitz93** [reply]
@patrickunterwegs 
https://github.com/TechbeeAT/jtxBoard/blob/7a9fd794988237359931ab85500f9d479f1394e0/app/src/main/java/at/techbee/jtx/widgets/elements/ListEntry.kt#L54
```
 val intent = Intent(context, MainActivity2::class.java).apply {
        flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TASK
        action = MainActivity2.INTENT_ACTION_OPEN_ICALOBJECT
        putExtra(MainActivity2.INTENT_EXTRA_ITEM2SHOW, obj.id)
    }
```
could you move this inside the `Box` block
or write a lambda that builds the intent object.
I think this is where the issue lies with obj.id.

**Reply by patrickunterwegs** [reply]
Well, this one actually works without problems. This is the intent that opens a specific entry. (Or did you have issues with that as well?)

The problematic intent is here: 
https://github.com/TechbeeAT/jtxBoard/blob/7a9fd794988237359931ab85500f9d479f1394e0/app/src/main/java/at/techbee/jtx/widgets/ListWidgetContent.kt#L51

```
    val addNewIntent = Intent(context, MainActivity2::class.java).apply {
        flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TASK
        action = when (listWidgetConfig.module) {
            Module.JOURNAL -> MainActivity2.INTENT_ACTION_ADD_JOURNAL
            Module.NOTE -> MainActivity2.INTENT_ACTION_ADD_NOTE
            Module.TODO -> MainActivity2.INTENT_ACTION_ADD_TODO
        }
        listWidgetConfig.searchCollection.firstOrNull()?.let {
            putExtra(MainActivity2.INTENT_EXTRA_COLLECTION2PRESELECT, it)
        }
    }
```

But here you can see that the intent just looks the same way and is also called the same way. 
I have already payed around with just building the intent directly in 

```
            Image(
                provider = ImageProvider(R.drawable.ic_widget_add),
                contentDescription = context.getString(R.string.add),
                modifier = GlanceModifier
                    .clickable(actionStartActivity(addNewIntent))
                    .padding(8.dp)
                    .size(imageSize)
            )
```

But it didn't make a difference. The main trouble that I have with that is that the first intent (your code snippet) always works, the second one from my answer doesn't...
